### PR TITLE
Add native support for the MediaText block

### PR DESCRIPTION
**Pull request by Tug**
Addresses https://github.com/wordpress-mobile/gutenberg-mobile/issues/689

## Description
This PR adds native support for the MediaText block thus enabling native inner block support.
This is very much like an early work and probably still needs a lot of work

## How has this been tested?
Tested with gutenberg mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1258

## Types of changes
Native block support

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/components/src/icon/index.js [hunk 1]
```diff
@@ -17,6 +17,10 @@ function Icon( { icon = null, size, ...additionalProps } ) {
 		return <Dashicon icon={ icon } size={ iconSize } { ...additionalProps } />;
 	}
 
+	if ( Dashicon === icon.type ) {
```
Comment (gziolo) [inline packages/components/src/icon/index.js]: This `Icon` vs `Dashicon` usage is confusing, but this should work and make it simpler in the future 👍

Comment (Tug) [reply packages/components/src/icon/index.js]: Yeah, still need to think how to make it work nicely with css classes, currently I'm just copying the classes of the Button parent into the Icon so we can [customize it at the svg primitive level](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/primitives/svg/index.native.js#L20), which is not ideal.

**Review by pinarol — APPROVED** [review]
LGTM 🎉 
I think we should merge this as is and handle the missing functionality on separate PRs. This one become huge anyway and you already provided a good first iteration of InnerBlocks which was the main purpose of this PR.

**Comment by pinarol** [comment]
Pushed this [commit](https://github.com/WordPress/gutenberg/pull/16305/commits/f91fd8a0dadc66a09bdf35bf00f28c74812af148) to fix the below issue:

mediaText-scroll


It was scrolling unnecessarily. I tried to set autoScroll=true on editor level only, no need for inner scrollviews to scroll. Please give a look at that commit before merging.

File: packages/components/src/icon/index.js (hunk 1 above)
Comment (etoledom) [reply packages/components/src/icon/index.js]: This change was messing up with many unit tests, so I partially rolled it back.
I understand the idea of the change but I thought it was better to comply with the tests.

**Comment by paaljoachim** [comment]
Is this PR adding inner blocks to the existing Media & Text block?
If so then that would be very helpful as a user would then be able to add a link to the media.
Similar to: https://github.com/WordPress/gutenberg/issues/15422
If it is something else then I have just misread.

**Comment by etoledom** [comment]
Fixed unit tests :yaaay:

The `BlockVerticalAlignmentToolbar` still work and is stilled as expected, but I messed up a bit the style of the `Add block (+)` button. I'll check that out next.

A question: Is media uploads considered for this PR or can it be handled on a separated one? 
cc @pinarol

**Comment by etoledom** [comment]
@paaljoachim - This PR implements Inner Blocks (and MediaText block) on Mobile Gutenberg. It does not modify any behavior of Gutenberg Web.

**Comment by pinarol** [comment]
> A question: Is media uploads considered for this PR or can it be handled on a separated one?
cc @pinarol

It will be handled separately with [this issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1313)

File: packages/components/src/icon/index.js (hunk 1 above)
Comment (etoledom) [reply packages/components/src/icon/index.js]: Nvm. I found out why the test errors were so spread out. It was some import error.
Rolled back to keep these changes fixing those unit tests.

File: packages/edit-post/src/components/header/plugin-more-menu-item/test/__snapshots__/index.js.snap [hunk 2]
```diff
@@ -23,7 +23,7 @@ exports[`PluginMoreMenuItem renders menu item as button properly 1`] = `
     >
       <svg
         aria-hidden="true"
-        className="dashicon dashicons-smiley"
+        className="dashicon dashicons-smiley components-icon-button components-menu-item__button has-icon has-text"
```
Comment (gziolo) [inline packages/edit-post/src/components/header/plugin-more-menu-item/test/__snapshots__/index.js.snap]: This is a bug which needs to be fixed. All those class names shouldn't be passed down. They are already applied to the button.

**Comment by pinarol** [comment]
We are almost ready to merge on the mobile side but this is effecting both platforms so it'd be really good if this is reviewed from a web perspective as well. Could someone from the current reviewers list take a look at this please?

We have some recently started work that uses InnerBlocks so this can become a blocker soon.

**Comment by pinarol** [comment]
@gziolo Just saw that you are reviewing, thanks a lot!

**Comment by gziolo** [comment]
Icons on web are broken with the proposed changes for the `IconButton` component:
Screen Shot 2019-08-29 at 11 23 16

**Comment by gziolo** [comment]
> @gziolo Just saw that you are reviewing, thanks a lot!

Yes, we'are trying to find an alternative way to pass all required states to SVG icons to convert them into corresponding styles to make it work for web with CSS which by design cascades styles.

**Comment by gziolo** [comment]
This fixes everything for the web:
```diff
diff --git a/packages/components/src/icon-button/index.js b/packages/components/src/icon-button/index.js
index 1c922ed0f..377feab76 100644
--- a/packages/components/src/icon-button/index.js
+++ b/packages/components/src/icon-button/index.js
@@ -56,7 +56,7 @@ function IconButton( props, ref ) {
 			className={ classes }
 			ref={ ref }
 		>
-			<Icon icon={ icon } ariaPressed={ ariaPressed } className={ classes } />
+			<Icon icon={ icon } ariaPressed={ ariaPressed } />
 			{ children }
 		</Button>
 	);
diff --git a/packages/components/src/icon/index.js b/packages/components/src/icon/index.js
index e7333bd65..5e762d1d5 100644
--- a/packages/components/src/icon/index.js
+++ b/packages/components/src/icon/index.js
@@ -10,21 +10,22 @@ import Dashicon from '../dashicon';
 import { SVG } from '../primitives';
 
 function Icon( { icon = null, size, ...additionalProps } ) {
-	let iconSize;
+	// Dashicons should be 20x20 by default.
+	const dashiconSize = size || 20;
 
 	if ( 'string' === typeof icon ) {
-		// Dashicons should be 20x20 by default
-		iconSize = size || 20;
-		return <Dashicon icon={ icon } size={ iconSize } { ...additionalProps } />;
+		return <Dashicon icon={ icon } size={ dashiconSize } { ...additionalProps } />;
 	}
 
 	if ( icon && Dashicon === icon.type ) {
-		return icon;
+		return cloneElement( icon, {
+			size: dashiconSize,
+			...additionalProps,
+		} );
 	}
 
-	// Any other icons should be 24x24 by default
-	iconSize = size || 24;
-
+	// Icons should be 24x24 by default.
+	const iconSize = size || 24;
 	if ( 'function' === typeof icon ) {
 		if ( icon.prototype instanceof Component ) {
 			return createElement( icon, { size: iconSize, ...additionalProps } );
diff --git a/packages/components/src/icon/test/index.js b/packages/components/src/icon/test/index.js
index 053b0cf39..a645568c1 100644
--- a/packages/components/src/icon/test/index.js
+++ b/packages/components/src/icon/test/index.js
@@ -11,6 +11,7 @@ import { Component } from '@wordpress/element';
 /**
  * Internal dependencies
  */
+import Dashicon from '../../dashicon';
 import Icon from '../';
 import { Path, SVG } from '../../';
 
@@ -31,12 +32,18 @@ describe( 'Icon', () => {
 		expect( wrapper.find( 'Dashicon' ).prop( 'icon' ) ).toBe( 'format-image' );
 	} );
 
-	it( 'renders a dashicon and with a default size of 20', () => {
+	it( 'renders a dashicon by slug and with a default size of 20', () => {
 		const wrapper = shallow( <Icon icon="format-image" /> );
 
 		expect( wrapper.find( 'Dashicon' ).prop( 'size' ) ).toBe( 20 );
 	} );
 
+	it( 'renders a dashicon by element and with a default size of 20', () => {
+		const wrapper = shallow( <Icon icon={ <Dashicon icon="format-image" /> } /> );
+
+		expect( wrapper.find( 'Dashicon' ).prop( 'size' ) ).toBe( 20 );
+	} );
+
 	it( 'renders a function', () => {
 		const wrapper = shallow( <Icon icon={ () => <span /> } /> );
 
@@ -98,6 +105,7 @@ describe( 'Icon', () => {
 
 		describe.each( [
 			[ 'dashicon', { icon: 'format-image' } ],
+			[ 'dashicon element', { icon: <Dashicon icon="format-image" /> } ],
 			[ 'element', { icon: <span /> } ],
 			[ 'svg element', { icon: svg } ],
 			[ 'component', { icon: MyComponent } ],
```

I didn't update snapshots as we still need to find a way to propagate button states to SVGs.

**Review by gziolo — APPROVED** [review]
I re-tested in WordPress with the latest changes applied and it all works perfectly fine. At least I haven't noticed anything happening to the icons.

**Comment by etoledom** [comment]
Thank you so much @gziolo for your help!
Waiting for travis to finish this 😁

**Review by pinarol — APPROVED** [review]
Tested on WPiOS app again and this is working good 👍

**Comment by pinarol** [comment]
Thank you very much @gziolo  🎉

**Pull request merged**